examples/src/main/java/org/apache/ignite/examples/ml/preprocessing/encoding/TargetEncoderExample.java

```diff
@@ -0,0 +1,138 @@
+
+/**
+ * Example that shows how to use Target Encoder preprocessor to encode labels presented as a mean target value.
+ * <p>
+ * Code in this example launches Ignite grid and fills the cache with test data (based on mushrooms dataset).</p>
+ * <p>
+ * After that it defines preprocessors that extract features from an upstream data and encode category with avarage
+ * target value (categories). </p>
+ * <p>
+ * Then, it trains the model based on the processed data using gradient boosing decision tree classification.</p>
+ * <p>
+ * Finally, this example uses {@link Evaluator} functionality to compute metrics from predictions.</p>
+ *
+ * <p>Daniele Miccii-Barreca (2001). A Preprocessing Scheme for High-Cardinality Categorical
+ * Attributes in Classification and Prediction Problems. SIGKDD Explor. Newsl. 3, 1.
+ * From http://dx.doi.org/10.1145/507533.507538</p>
+ */
+public class TargetEncoderExample {
+    /**
+     * Run example.
+     */
+    public static void main(String[] args) {
+        System.out.println();
+        System.out.println(">>> Train Gradient Boosing Decision Tree model on amazon-employee-access-challenge_train.csv dataset.");
+
+        try (Ignite ignite = Ignition.start("examples/config/example-ignite.xml")) {
+            try {
+                IgniteCache<Integer, Object[]> dataCache = new SandboxMLCache(ignite)
+                    .fillObjectCacheWithCategoricalData(MLSandboxDatasets.AMAZON_EMPLOYEE_ACCESS);
+
+                Set<Integer> featuresIndexies = new HashSet<>(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9));
+                Set<Integer> targetEncodedfeaturesIndexies = new HashSet<>(Arrays.asList(1, 5, 6));
+                Integer targetIndex = 0;
+
+                final Vectorizer<Integer, Object[], Integer, Object> vectorizer = new ObjectArrayVectorizer<Integer>(featuresIndexies.toArray(new Integer[0]))
+                    .labeled(targetIndex);
+
+                Preprocessor<Integer, Object[]> strEncoderPreprocessor = new EncoderTrainer<Integer, Object[]>()
+                    .withEncoderType(EncoderType.STRING_ENCODER)
+                    .withEncodedFeature(0)
+                    .withEncodedFeatures(featuresIndexies)
+                    .fit(ignite,
+                        dataCache,
+                        vectorizer
+                    );
+
+                Preprocessor<Integer, Object[]> targetEncoderProcessor = new EncoderTrainer<Integer, Object[]>()
+                    .withEncoderType(EncoderType.TARGET_ENCODER)
+                    .labeled(0)
+                    .withEncodedFeatures(targetEncodedfeaturesIndexies)
+                    .minSamplesLeaf(1)
+                    .minCategorySize(1L)
+                    .smoothing(1d)
+                    .fit(ignite,
+                        dataCache,
+                        strEncoderPreprocessor
+                    );
+
+                Preprocessor<Integer, Object[]> lbEncoderPreprocessor = new EncoderTrainer<Integer, Object[]>()
```

Comment:
Sorry, but I didn't understand this pipeline? Why those 3 encoders are combined here? Could they work only in this combination?
In my opinion, user have a choice what to do with Strings, but he should choose one method (not the chain of methods).

Please share your vision here

Reply:
I think that we want use `EncoderType.TARGET_ENCODER` only for a few columns (may be only one). In this example I use `EncoderType.STRING_ENCODER` as general propose encoder and `EncoderType.TARGET_ENCODER` for special one.